### ISSUE TITLE
Fix for 'previous definition of xDrop was here' breaks Liquid.

### ISSUE THREAD

**Issue by virtadpt** [cantino/huginn#783]
Immediately after updating to a commit of Huginn that incorporated the 'previous definition of xDrop was here' fix (commit 3bb7f7a06f5cc8d5ffa47a689d7f3387f9284823) Liquid templating broke across all Agents in all Scenarios.  Manually reverting to commit a6dbb9c787e0c814921ef7001491346b148d35e9 which preceeded that fix repaired Liquid templating functionality.  Please re-examine patch.

**Comment by cantino** [comment]
You definitely restarted Huginn after checking out that version?

**Comment by cantino** [comment]
(I'm trying it now.)

**Comment by cantino** [comment]
I just updated my Huginn installation to the newest version and Liquid is still working.  That's VERY odd.

**Comment by virtadpt** [comment]
I did - it was running in Screen so I ^C'd it, then went through the process list to kill the other processes still hanging around before checking out that commit and trying again.  Events injected with the ManualAgent set off TriggerAgent as expected, and then forcing a run of the agent networks that fed them started resulting in events being transmitted.

**Comment by cantino** [comment]
@dsander you haven't seen anything like this, right?

**Comment by dsander** [comment]
@cantino No, but maybe the liquid formatting we use is too basic?

@virtadpt Do you have an agent/scenario that fails and you can share?

**Comment by virtadpt** [comment]
I can share any of my scenarios - they all broke.  Let me pick out part of one for you...
```
Type: Twitter User Agent
{
  "username": "insoc",
  "include_retweets": "true",
  "expected_update_period_in_days": "365"
}

Type: Twitter User Agent 
{
  "username": "priyom_org",
  "include_retweets": "true",
  "expected_update_period_in_days": "3650"
}

Type: Twitter User Agent 
{
  "username": "thecruxshadows",
  "include_retweets": "true",
  "expected_update_period_in_days": "3650"
}

Type: Jabber Agent
{
  "jabber_server": "{% credential primary_xmpp_server %}",
  "jabber_port": "5222",
  "jabber_sender": "{% credential primary_xmpp_login %}",
  "jabber_receiver": "{% credential primary_xmpp_destination %}",
  "jabber_password": "{% credential primary_xmpp_password %}",
  "message": "Twitter alert from Lifeline: @{{user.screen_name}} - {{text}}",
  "expected_receive_period_in_days": "365"
}
```

**Comment by dsander** [comment]
Hmm that is working without issues on my installation, I have a Twitter agent sending my a message via XMPP. Do you get failed jobs or an exception of any kind?

**Comment by virtadpt** [comment]
Neither.  Just empty alerts via XMPP of the form
```
Twitter alert from Lifeline: @ - 
```

**Comment by cantino** [comment]
That is SO strange.  And nothing in the logs?

**Comment by virtadpt** [comment]
Not that I can tell.  I can run some tests on Saturday, depending on how things go this morning...

**Comment by virtadpt** [comment]
I re-ran my tests - there is nothing in the logs that would suggest that an agent was throwing exceptions or anything like that.  I also updated my checkout to the latest commit from master/ and the same thing happened - Liquid templating stopped working, even when I injected crafted events into agent networks with ManualAgent.  I then stepped backwards in commits and tested each time, and it wasn't until I got back to commit a6dbb9c787e0c814921ef7001491346b148d35e9 that Liquid templating started working again.

**Comment by dsander** [comment]
@virtadpt I finally got the same bug! `Twitter alert from Lifeline: @ -` the strange thing is the events look almost the same: [working](http://pastie.org/private/ynttnjmt6nof8tsyjvmtg), [not working](http://pastie.org/private/ggvc6nnidoapt4094g1qda)

**Comment by virtadpt** [comment]
Yes!  That's what I'm seeing over here!

**Comment by virtadpt** [comment]
Just tested.  Still happens on commit 55cb1d00d0574a5464ac183ee411b4aee297bdb4.

**Comment by virtadpt** [comment]
Factoid - the event gets written into the database in a broken fashion:
```
19:29:58 web.1  |   SQL (0.7ms)  INSERT INTO `events` (`user_id`, `expires_at`, `payload`,
`agent_id`, `created_at`, `updated_at`) VALUES (1, '2015-04-30 19:29:58',
'{\"message\":\"The number of people listening to the BART Police Dispatch radio network
has peaked to !  You may wish to listen to it here: \\n\\n
http://www.radioreference.com/scripts/playlists/1/2776/0-5708229104.mp3\",
\"peak\":9.0,\"peak_time\":1429816768.0,\"grouped_by\":\"no_group\"}', 30, '2015-04-23
19:29:58', '2015-04-23 19:29:58')
```

**Comment by dsander** [comment]
@virtadpt you are right, 3bb7f7a is the cause for the issues you are having. I am working on a solution that both works in development and production.

As a workaround you can run in production or set `config.cache_classes = true` in `config/environments/development.rb`

**Comment by virtadpt** [comment]
Awesome!
When you're done, mind if I pick your brain on Gitter about how that breaks Liquid templating?  Is it because it's already been pulled in ("class Drop < Liquid::Drop") and pulling it in again puts it into an inconsistent internal state?

**Comment by dsander** [comment]
To be honest, I have no idea why it broke stuff.  3bb7f7a was created to fix the 'previous definition of xDrop was here'  warnings which happened when you changed files in development and rails (tried?) to reload the classes.

I noticed the same issue you are having yesterday night for the first time and was able to reproduce it today by re-emitting an event twice after restarting huginn. I guess the first time loaded the required liquid classes and the second time something was reloaded (due to `cache_classes` being set to false in development), but because `AgentDrop` wasn't redefined the liquid interpolating failed. Why there wasn't any exception is another mystery to me.

Maybe @knu has more insight why #770 broke liquid templating in development?

**Comment by dsander** [comment]
Oh by the way, you could try #771 which should fix the issue.

**Comment by virtadpt** [comment]
After another scheduler crash last night, I'm trying it on a new checkout right now.

**Comment by virtadpt** [comment]
Applying the suggested changes in #771 while running HEAD on a new checkout definitely fixed Liquid.

**Comment by dsander** [comment]
Fixed by cfe8482c0d6eca95e7324aa266caacccc4ffd46f

**Comment by virtadpt** [comment]
Looks good!
